### DISCUSSION TITLE
Hole punching

### DISCUSSION THREAD

**Discussion by kehiy** [libp2p/go-libp2p#2556] [Q&A] [answered]
How can we enable `Hole punching` in our `go-libp2p` app and also test it?

**Comment by sukunrt** [comment]
Use the `EnableHolePunching` option: https://github.com/libp2p/go-libp2p/blob/master/options.go#L536

What do you mean by testing it? 
There's a grafana dashboard available to monitor hole punch success rates. https://github.com/libp2p/go-libp2p/blob/master/dashboards/holepunch/holepunch.json

**Reply by kehiy** [reply]
Do you mean I need to set `EnableHolePunching` to true just?

about testing, I meant `unit_test` in the code.

**Reply by sukunrt** [reply, marked as answer]
you can do `libp2p.New(EnableHolePunching(), otherOpts...)`

Within libp2p, we have a bunch of tests that ensure that hole punching is doing what we want it to. These don't exactly simulate nodes behind NAT/firewall but mock only enough that we have high confidence that we're doing the right thing in most cases. 

For really testing NAT traversal in unit tests you'll need to do a lot more work. 
There is a draft PR that introduces such tests for hole punching. https://github.com/libp2p/go-libp2p/pull/2443 but I have no plans right now to polish it and make it reviewable. I do not see myself getting back to it for the next month. However you can see how I use the tailscale natlab package to configure hosts that appear to be behind a NAT. You can do something similar for your use case. 

For end to end tests, see the discussion here for progress on end to end hole punching tests between different libp2p implementations: https://github.com/libp2p/test-plans/issues/116

**Reply by kehiy** [reply]
@sukunrt, thank you so much!
